Intermittent CI Failures

> FAILED tests/test_awslambda/test_lambda_eventsourcemapping.py::test_invoke_function_from_sqs_exception - botocore.errorfactory.EntityAlreadyExistsException: An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name my-role already exists.

Looks like this started happening after the Server Mode tests for `awslambda` began running in parallel (#4368).

<details><summary>Detailed Log Output</summary>

=================================== FAILURES ===================================
___________________ test_invoke_function_from_sqs_exception ____________________
[gw1] linux -- Python 3.8.14 /opt/hostedtoolcache/Python/3.8.14/x64/bin/python

    def get_role_name():
        with mock_iam():
            iam = boto3.client("iam", region_name=_lambda_region)
            try:
>               return iam.get_role(RoleName="my-role")["Role"]["Arn"]

tests/test_awslambda/utilities.py:170: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.client.IAM object at 0x7f279c4a5eb0>, args = ()
kwargs = {'RoleName': 'my-role'}

    def _api_call(self, *args, **kwargs):
        # We're accepting *args so that we can give a more helpful
        # error message than TypeError: _api_call takes exactly
        # 1 argument.
        if args:
            raise TypeError(
                f"{py_operation_name}() only accepts keyword arguments."
            )
        # The "self" in this scope is referring to the BaseClient.
>       return self._make_api_call(operation_name, kwargs)

/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/botocore/client.py:495: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.client.IAM object at 0x7f279c4a5eb0>
operation_name = 'GetRole', api_params = {'RoleName': 'my-role'}

    def _make_api_call(self, operation_name, api_params):
        operation_model = self._service_model.operation_model(operation_name)
        service_name = self._service_model.service_name
        history_recorder.record(
            'API_CALL',
            {
                'service': service_name,
                'operation': operation_name,
                'params': api_params,
            },
        )
        if operation_model.deprecated:
            logger.debug(
                'Warning: %s.%s() is deprecated', service_name, operation_name
            )
        request_context = {
            'client_region': self.meta.region_name,
            'client_config': self.meta.config,
            'has_streaming_input': operation_model.has_streaming_input,
            'auth_type': operation_model.auth_type,
        }
        endpoint_url, additional_headers = self._resolve_endpoint_ruleset(
            operation_model, api_params, request_context
        )
        request_dict = self._convert_to_request_dict(
            api_params=api_params,
            operation_model=operation_model,
            endpoint_url=endpoint_url,
            context=request_context,
            headers=additional_headers,
        )
        resolve_checksum_context(request_dict, operation_model, api_params)
    
        service_id = self._service_model.service_id.hyphenize()
        handler, event_response = self.meta.events.emit_until_response(
            'before-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            model=operation_model,
            params=request_dict,
            request_signer=self._request_signer,
            context=request_context,
        )
    
        if event_response is not None:
            http, parsed_response = event_response
        else:
            apply_request_checksum(request_dict)
            http, parsed_response = self._make_request(
                operation_model, request_dict, request_context
            )
    
        self.meta.events.emit(
            'after-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            http_response=http,
            parsed=parsed_response,
            model=operation_model,
            context=request_context,
        )
    
        if http.status_code >= 300:
            error_code = parsed_response.get("Error", {}).get("Code")
            error_class = self.exceptions.from_code(error_code)
>           raise error_class(parsed_response, operation_name)
E           botocore.errorfactory.NoSuchEntityException: An error occurred (NoSuchEntity) when calling the GetRole operation: Role my-role not found

/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/botocore/client.py:914: NoSuchEntityException

During handling of the above exception, another exception occurred:

    @pytest.mark.network
    @mock_logs
    @mock_lambda
    @mock_sqs
    def test_invoke_function_from_sqs_exception():
        function_name = str(uuid4())[0:6]
        logs_conn = boto3.client("logs", region_name="us-east-1")
        sqs = boto3.resource("sqs", region_name="us-east-1")
        queue = sqs.create_queue(QueueName=f"{function_name}_queue")
    
        conn = boto3.client("lambda", region_name="us-east-1")
        func = conn.create_function(
            FunctionName=function_name,
            Runtime="python2.7",
>           Role=get_role_name(),
            Handler="lambda_function.lambda_handler",
            Code={"ZipFile": get_test_zip_file_error()},
            Description="test lambda function",
            Timeout=3,
            MemorySize=128,
            Publish=True,
        )

tests/test_awslambda/test_lambda_eventsourcemapping.py:223: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_awslambda/utilities.py:172: in get_role_name
    return iam.create_role(
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/botocore/client.py:495: in _api_call
    return self._make_api_call(operation_name, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.client.IAM object at 0x7f279c4a5eb0>
operation_name = 'CreateRole'
api_params = {'AssumeRolePolicyDocument': 'some policy', 'Path': '/my-path/', 'RoleName': 'my-role'}

    def _make_api_call(self, operation_name, api_params):
        operation_model = self._service_model.operation_model(operation_name)
        service_name = self._service_model.service_name
        history_recorder.record(
            'API_CALL',
            {
                'service': service_name,
                'operation': operation_name,
                'params': api_params,
            },
        )
        if operation_model.deprecated:
            logger.debug(
                'Warning: %s.%s() is deprecated', service_name, operation_name
            )
        request_context = {
            'client_region': self.meta.region_name,
            'client_config': self.meta.config,
            'has_streaming_input': operation_model.has_streaming_input,
            'auth_type': operation_model.auth_type,
        }
        endpoint_url, additional_headers = self._resolve_endpoint_ruleset(
            operation_model, api_params, request_context
        )
        request_dict = self._convert_to_request_dict(
            api_params=api_params,
            operation_model=operation_model,
            endpoint_url=endpoint_url,
            context=request_context,
            headers=additional_headers,
        )
        resolve_checksum_context(request_dict, operation_model, api_params)
    
        service_id = self._service_model.service_id.hyphenize()
        handler, event_response = self.meta.events.emit_until_response(
            'before-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            model=operation_model,
            params=request_dict,
            request_signer=self._request_signer,
            context=request_context,
        )
    
        if event_response is not None:
            http, parsed_response = event_response
        else:
            apply_request_checksum(request_dict)
            http, parsed_response = self._make_request(
                operation_model, request_dict, request_context
            )
    
        self.meta.events.emit(
            'after-call.{service_id}.{operation_name}'.format(
                service_id=service_id, operation_name=operation_name
            ),
            http_response=http,
            parsed=parsed_response,
            model=operation_model,
            context=request_context,
        )
    
        if http.status_code >= 300:
            error_code = parsed_response.get("Error", {}).get("Code")
            error_class = self.exceptions.from_code(error_code)
>           raise error_class(parsed_response, operation_name)
E           botocore.errorfactory.EntityAlreadyExistsException: An error occurred (EntityAlreadyExists) when calling the CreateRole operation: Role with name my-role already exists.

</details>

--- a/tests/test_awslambda/test_lambda.py
+++ b/tests/test_awslambda/test_lambda.py
@@ -1134,3 +1134,39 @@ def test_multiple_qualifiers():
     fn["FunctionArn"].should.equal(
         f"arn:aws:lambda:us-east-1:{ACCOUNT_ID}:function:{fn_name}:6"
     )
+
+
+def test_get_role_name_utility_race_condition():
+    # Play with these variables as needed to reproduce the error.
+    max_workers, num_threads = 3, 15
+
+    errors = []
+    roles = []
+
+    def thread_function(_):
+        while True:
+            # noinspection PyBroadException
+            try:
+                role = get_role_name()
+            except ClientError as e:
+                errors.append(str(e))
+                break
+            except Exception:
+                # boto3 and our own IAMBackend are not thread-safe,
+                # and occasionally throw weird errors, so we just
+                # pass and retry.
+                # https://github.com/boto/boto3/issues/1592
+                pass
+            else:
+                roles.append(role)
+                break
+
+    import concurrent.futures
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        executor.map(thread_function, range(num_threads))
+    # Check all threads are accounted for, all roles are the same entity,
+    # and there are no client errors.
+    assert len(errors) + len(roles) == num_threads
+    assert roles.count(roles[0]) == len(roles)
+    assert len(errors) == 0

--- a/tests/test_awslambda/utilities.py
+++ b/tests/test_awslambda/utilities.py
@@ -166,14 +166,18 @@ def create_invalid_lambda(role):
 def get_role_name():
     with mock_iam():
         iam = boto3.client("iam", region_name=_lambda_region)
-        try:
-            return iam.get_role(RoleName="my-role")["Role"]["Arn"]
-        except ClientError:
-            return iam.create_role(
-                RoleName="my-role",
-                AssumeRolePolicyDocument="some policy",
-                Path="/my-path/",
-            )["Role"]["Arn"]
+        while True:
+            try:
+                return iam.get_role(RoleName="my-role")["Role"]["Arn"]
+            except ClientError:
+                try:
+                    return iam.create_role(
+                        RoleName="my-role",
+                        AssumeRolePolicyDocument="some policy",
+                        Path="/my-path/",
+                    )["Role"]["Arn"]
+                except ClientError:
+                    pass
 
 
 def wait_for_log_msg(expected_msg, log_group, wait_time=30):

